spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/ChangelogIterator.java

```diff
@@ -0,0 +1,162 @@
+ *
+ * <p>The iterator marks the delete-row and insert-row to be the update-rows. For example, these two
+ * rows
+ *
+ * <ul>
+ *   <li>(id=1, data='a', op='DELETE')
+ *   <li>(id=1, data='b', op='INSERT')
+ * </ul>
+ *
+ * will be marked as update-rows:
+ *
+ * <ul>
+ *   <li>(id=1, data='a', op='UPDATE_BEFORE')
+ *   <li>(id=1, data='b', op='UPDATE_AFTER')
+ * </ul>
+ */
+public class ChangelogIterator implements Iterator<Row>, Serializable {
+  private static final String DELETE = ChangelogOperation.DELETE.name();
+  private static final String INSERT = ChangelogOperation.INSERT.name();
+  private static final String UPDATE_BEFORE = ChangelogOperation.UPDATE_BEFORE.name();
+  private static final String UPDATE_AFTER = ChangelogOperation.UPDATE_AFTER.name();
+
+  private final Iterator<Row> rowIterator;
+  private final int changeTypeIndex;
+  private final List<Integer> partitionIdx;
+
+  private Row cachedRow = null;
+
+  public ChangelogIterator(
+      Iterator<Row> rowIterator, int changeTypeIndex, List<Integer> partitionIdx) {
+    this.rowIterator = rowIterator;
+    this.changeTypeIndex = changeTypeIndex;
+    this.partitionIdx = partitionIdx;
+  }
+
+  @Override
+  public boolean hasNext() {
+    if (cachedRow != null) {
+      return true;
+    }
+    return rowIterator.hasNext();
+  }
+
+  @Override
+  public Row next() {
+    // if there is an updated cached row, return it directly
+    if (updated(cachedRow)) {
+      Row row = cachedRow;
+      cachedRow = null;
+      return row;
+    }
+
+    Row currentRow = currentRow();
+
+    if (rowIterator.hasNext()) {
+      GenericRowWithSchema nextRow = (GenericRowWithSchema) rowIterator.next();
+      cachedRow = nextRow;
+
+      if (updateOrCarryoverRecord(currentRow, nextRow)) {
+        Row[] rows = update((GenericRowWithSchema) currentRow, nextRow);
+
+        currentRow = rows[0];
+        cachedRow = rows[1];
+      }
+    }
+
+    return currentRow;
+  }
+
+  private Row[] update(GenericRowWithSchema currentRow, GenericRowWithSchema nextRow) {
+    GenericInternalRow deletedRow = new GenericInternalRow(currentRow.values());
+    GenericInternalRow insertedRow = new GenericInternalRow(nextRow.values());
+
+    if (isCarryoverRecord(deletedRow, insertedRow)) {
+      // set carry-over rows to null for filtering out later
+      return new Row[] {null, null};
+    } else {
+      deletedRow.update(changeTypeIndex, UPDATE_BEFORE);
+      insertedRow.update(changeTypeIndex, UPDATE_AFTER);
+
+      return new Row[] {
+        RowFactory.create(deletedRow.values()), RowFactory.create(insertedRow.values())
+      };
+    }
+  }
+
+  private boolean isCarryoverRecord(GenericInternalRow deletedRow, GenericInternalRow insertedRow) {
+    // set the change_type to the same value
+    deletedRow.update(changeTypeIndex, "");
+    insertedRow.update(changeTypeIndex, "");
+    return deletedRow.equals(insertedRow);
+  }
+
+  private boolean updated(Row row) {
```

Comment:
Also one more, variable may make more sense as cachedRow or cachedChangelog.  Have to always remember that this method is only for checking the cache.

Reply:
I agree, I was thinking whether or not it should be like cachedUpdateRecord or something like that

So we have "rows" and "records", rows are the input and Records are the output, even though they are all actually "Row" type

Reply:
Made the change.